source/braille.py

```diff
@@ -2285,97 +2299,46 @@ def _disableDetection(self):
 			self._detector = None
 		self._detectionEnabled = False
 
-class _BgThread:
-	"""A singleton background thread used for background writes and raw braille display I/O.
-	"""
-
-	thread = None
-	exit = False
-	queuedWrite = None
-
-	@classmethod
-	def start(cls):
-		if cls.thread:
-			return
-		cls.queuedWriteLock = threading.Lock()
-		thread = cls.thread = threading.Thread(
-			name=f"{cls.__module__}.{cls.__qualname__}",
-			target=cls.func
-		)
-		thread.daemon = True
-		thread.start()
-		cls.handle = ctypes.windll.kernel32.OpenThread(winKernel.THREAD_SET_CONTEXT, False, thread.ident)
-		cls.ackTimerHandle = winKernel.createWaitableTimer()
-
-	@classmethod
-	def queueApc(cls, func, param=0):
-		# Ensure the thread is running
-		cls.start()
-		ctypes.windll.kernel32.QueueUserAPC(func, cls.handle, param)
-
-	@classmethod
-	def stop(cls, timeout=None):
-		if not cls.thread:
-			return
-		cls.exit = True
-		if not ctypes.windll.kernel32.CancelWaitableTimer(cls.ackTimerHandle):
-			raise ctypes.WinError()
-		winKernel.closeHandle(cls.ackTimerHandle)
-		cls.ackTimerHandle = None
-		# Wake up the thread. It will exit when it sees exit is True.
-		cls.queueApc(cls.executor)
-		cls.thread.join(timeout)
-		cls.exit = False
-		winKernel.closeHandle(cls.handle)
-		cls.handle = None
-		cls.thread = None
-
-	@winKernel.PAPCFUNC
-	def executor(param):
-		if _BgThread.exit:
-			# func will see this and exit.
-			return
-		if not handler.display:
-			# Sometimes, the executor is triggered when a display is not fully initialized.
+	def _bgThreadExecutor(self, param: int):
+		"""Executed as APC when cells have to be written to a display asynchronously.
+		"""
+		if not self.display:
+			# Sometimes, the bg thread executor is triggered when a display is not fully initialized.
 			# For example, this happens when handling an ACK during initialisation.
 			# We can safely ignore this.
 			return
-		if handler.display._awaitingAck:
+		if self.display._awaitingAck:
 			# Do not write cells when we are awaiting an ACK
 			return
-		with _BgThread.queuedWriteLock:
-			data = _BgThread.queuedWrite
-			_BgThread.queuedWrite = None
+		with self.queuedWriteLock:
+			data: Optional[List[int]] = self.queuedWrite
+			self.queuedWrite = None
 		if not data:
 			return
 		try:
-			handler.display.display(data)
+			self.display.display(data)
 		except:
 			log.error("Error displaying cells. Disabling display", exc_info=True)
-			handler.handleDisplayUnavailable()
+			self.handleDisplayUnavailable()
 		else:
-			if handler.display.receivesAckPackets:
-				handler.display._awaitingAck = True
+			if self.display.receivesAckPackets:
+				self.display._awaitingAck = True
 				winKernel.setWaitableTimer(
-					_BgThread.ackTimerHandle,
-					int(handler.display.timeout*2000),
+					self.ackTimerHandle,
+					int(self.display.timeout * 2000),

```

Comment:
Do you know about this value? Why is it multiplied by 2000?

Reply:
It is converted to milli seconds and then multiplied by two. I added a descriptive comment.